runtime/io.c

```diff
@@ -196,13 +194,15 @@ CAMLexport struct channel * caml_open_descriptor_out(int fd)
 CAMLexport void caml_close_channel(struct channel *channel)
 {
   close(channel->fd);
-
+  caml_plat_lock (&caml_all_opened_channels_mutex);
   unlink_channel(channel);
     if (atomic_load_acq(&channel->refcount) > 0) {
     /* [caml_ml_out_channels_list] may have a reference to this channel. */
     link_channel (channel);
+    caml_plat_unlock (&caml_all_opened_channels_mutex);
     return;
```

Comment:
Do we still need this weird "unlock / oops, re-lock" routine if we're holding `caml_all_opened_channels_mutex`?  We cannot run concurrently with caml_ml_out_channels_list...

Reply:
Do you mean the `unlink`/re-`link` dance?

Reply:
Yes, unlink / relink, sorry for the typo.

Reply:
I have pushed a version which only unlinks when `caml_ml_out_channels_list` can no longer resurrect a dead channel.